tensorly/decomposition/candecomp_parafac.py

```diff
@@ -60,22 +48,77 @@ def parafac(tensor, rank, n_iter_max=100, init='svd', tol=10e-7,
                 factor[:, :tensor.shape[mode]] = U
                 U = T.tensor(factor)
             factors.append(U[:, :rank])
+    return factors
+
+
+def parafac(tensor, rank, n_iter_max=100, init='svd', tol=10e-7,
+            random_state=None, with_eigenvalues=False, verbose=False):
+    """CANDECOMP/PARAFAC decomposition via alternating least squares (ALS)
+
+    Computes a rank-`rank` decomposition of `tensor` [1]_ such that,
+
+        ``tensor = [| factors[0], ..., factors[-1] |]``.
 
+    If `with_eigenvalues == True` returns the decomposition such that,
+
+        ``tensor = [| lambda; factors[0], ..., factors[-1] |]``.
+
+    where `lambda` is an array of `rank` eigenvalues of the decomposition and
+    the columns of the factor matrices have unit norm.
+
+    Parameters
+    ----------
+    tensor : ndarray
+    rank  : int
+        Number of components
+    n_iter_max : int
+        Maximum number of iteration
+    init : {'svd', 'random'}, optional
+    tol : float, optional
+        Tolerance: the algorithm stops when the variation in the reconstruction
+        error is less than the tolerance.
+    random_state : {None, int, np.random.RandomState}
+    with_eigenvalues : bool
+        If `True`, normalizes the columns of the factor matrices and returns
+        the eigenvalues along with the list of factors.
+    verbose : int, optional
+        Level of verbosity
+
+    Returns
+    -------
+    factors : ndarray list
+        List of factors of the CP decomposition element `i` is of shape
+        (tensor.shape[i], rank)
+    eigenvalues : ndarray, optional
+        Array of length `rank`. Output when `with_eigenvalues == True`.
+        (See above.)
+
+    References
+    ----------
+    .. [1] T.G.Kolda and B.W.Bader, "Tensor Decompositions and Applications",
+       SIAM REVIEW, vol. 51, n. 3, pp. 455-500, 2009.
+
+    """
+    factors = initialize_factors(tensor, rank, init=init, random_state=random_state)
     rec_errors = []
     norm_tensor = T.norm(tensor, 2)
+    eigenvalues = None
 
     for iteration in range(n_iter_max):
         for mode in range(tensor.ndim):
             pseudo_inverse = T.tensor(np.ones((rank, rank)))
             for i, factor in enumerate(factors):
                 if i != mode:
-                    pseudo_inverse[:] = pseudo_inverse*T.dot(factor.T, factor)
+                    pseudo_inverse *= T.dot(factor.T, factor)
```

Comment:
What was the reasoning behind the use of `[:]`, by the way?

Reply:
In MXNet it is used for inplace operations.